src/main/java/com/powsybl/openloadflow/ac/outerloop/SecondaryVoltageControlOuterLoop.java

```diff
@@ -256,85 +217,93 @@ private static double calculatePilotPointDv(LfSecondaryVoltageControl secondaryV
-        boolean adjusted = false;
-
-        List<LfBus> controllerBuses = controlledBuses.stream()
-                .flatMap(controlledBus -> findVoltageControlEnabledControllerBuses(controlledBus).stream())
-                .toList();
+    private List<String> processSecondaryVoltageControl(List<LfSecondaryVoltageControl> secondaryVoltageControls,
+                                                        SensitivityContext sensitivityContext) {
+        List<String> adjustedZoneNames = new ArrayList<>();
 
-        double pilotDv = calculatePilotPointDv(secondaryVoltageControl);
-        KStats kStats = calculateKStats(controllerBuses);
-        if (Math.abs(pilotDv) > DV_EPS || !kStats.allAtLimits() && Math.abs(kStats.dkDiffMax()) > DK_DIFF_MAX_EPS) {
-            var pilotBus = secondaryVoltageControl.getPilotBus();
-            if (LOGGER.isDebugEnabled()) {
-                long allControllerBusesCount = controlledBuses.stream()
-                        .mapToLong(controlledBus -> findControllerBuses(controlledBus).size())
-                        .sum();
-                LOGGER.debug("Secondary voltage control of zone '{}': {}/{} controller buses available, pilot point dv is {} kV, controller buses dk diff max is {} (k average is {})",
-                        secondaryVoltageControl.getZoneName(), controllerBuses.size(), allControllerBusesCount, pilotDv * pilotBus.getNominalV(), kStats.dkDiffMax(), kStats.kAverage());
+        for (LfSecondaryVoltageControl secondaryVoltageControl : secondaryVoltageControls) {
+            double pilotDv = calculatePilotPointDv(secondaryVoltageControl);
+            List<LfBus> controllerBuses = secondaryVoltageControl.getControllerBuses();
+            KStats kStats = calculateKStats(controllerBuses);
+            if (Math.abs(pilotDv) > DV_EPS || !kStats.allAtLimits() && Math.abs(kStats.dkDiffMax()) > DK_DIFF_MAX_EPS) {
+                var pilotBus = secondaryVoltageControl.getPilotBus();
+                if (LOGGER.isDebugEnabled()) {
+                    int allControllerBusesCount = secondaryVoltageControl.getControllerBuses().size();
+                    LOGGER.debug("Secondary voltage control of zone '{}': {}/{} controller buses available, pilot point dv is {} kV, controller buses dk diff max is {} (k average is {})",
+                            secondaryVoltageControl.getZoneName(), controllerBuses.size(), allControllerBusesCount, pilotDv * pilotBus.getNominalV(), kStats.dkDiffMax(), kStats.kAverage());
+                }
+                adjustedZoneNames.add(secondaryVoltageControl.getZoneName());
             }
+        }
 
-            var controllerBusIndex = buildBusIndex(controllerBuses);
+        if (adjustedZoneNames.isEmpty()) {
+            return Collections.emptyList();
+        }
 
-            DenseMatrix a = createA(controllerBuses, controllerBusIndex);
+        List<LfBus> allControllerBuses = secondaryVoltageControls.stream()
```

Comment:
This is not clear why you leave that here... and why you don't put this method in the `LfSecondayVoltageControl` class.